### PR TITLE
Add support to encrypt messages without MAC truncation

### PR DESCRIPTION
**Pull request by poljar**

### CODE REVIEW

File: src/megolm/session_config.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+impl SessionConfig {
+    /// Get the numeric version of this `SessionConfig`.
+    pub fn version(&self) -> u8 {
+        self.version as u8
+    }
+
+    /// Create a `SessionConfig` for the Megolm version 1. This version of
+    /// Megolm will use AES-256 and HMAC with a truncated MAC to encrypt
+    /// individual messages. The MAC will be truncated to 8 bytes.
```
Comment (dkasak) [inline src/megolm/session_config.rs:38]:
```suggestion
    /// Create a `SessionConfig` for the Megolm version 1. This version of
    /// Megolm uses AES-256 and HMAC with a truncated MAC to encrypt individual
    /// messages. The MAC will be truncated to 8 bytes.
```

File: src/megolm/session_config.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    pub fn version_1() -> Self {
+        SessionConfig { version: Version::V1 }
+    }
+
+    /// Create a `SessionConfig` for the Megolm version 2. This version of
+    /// Megolm will use AES-256 and HMAC to encrypt individual messages. The
+    /// MAC won't be truncated.
```
Comment (dkasak) [inline src/megolm/session_config.rs:45]:
```suggestion
    /// Create a `SessionConfig` for the Megolm version 2. This version of
    /// Megolm uses AES-256 and HMAC to encrypt individual messages. The MAC
    /// won't be truncated.
```

File: src/olm/messages/message.rs [hunk 3]
```diff
@@ -167,26 +206,33 @@ impl TryFrom<&[u8]> for Message {
...
+                    .get(1..value.len() - mac_length)
+                    .ok_or(DecodeError::MessageTooShort(value.len()))?,
+            )?;
 
-            let mac_slice = &value[value.len() - Mac::TRUNCATED_LEN..];
+            let mac_slice = &value[value.len() - mac_length..];
 
-            if mac_slice.len() != Mac::TRUNCATED_LEN {
+            if mac_slice.len() != mac_length {
                 Err(DecodeError::InvalidMacLength(Mac::TRUNCATED_LEN, mac_slice.len()))
```
Comment (dkasak) [inline src/olm/messages/message.rs:227]:
```suggestion
                Err(DecodeError::InvalidMacLength(mac_length, mac_slice.len()))
```

**Review by dkasak — CHANGES_REQUESTED** [review]

File: src/megolm/group_session.rs [hunk 4]
```diff
@@ -72,19 +76,31 @@ impl GroupSession {
...
-            plaintext.as_ref(),
-        );
+        let message = match self.config.version {
+            Version::V1 => MegolmMessage::encrypt_truncated_mac(
+                self.message_index(),
+                &cipher,
+                &self.signing_key,
+                plaintext.as_ref(),
+            ),
+            Version::V2 => MegolmMessage::encrypt_private(
```
[the remark this answers is not in the document]
Comment (poljar) [reply src/megolm/group_session.rs]: https://github.com/matrix-org/vodozemac/pull/74/commits/002b7abb6f4c2c0031455836ff1ede6f66d2b281

**Review by dkasak — APPROVED** [review]

**Pull request merged**